Features/6 prototype projects viewer

DANIV18 opened the pull request:
Solving issue 6(Create Prototype Projects Viewer)

idaniv committed:
Was added ProjectViewerWindow class.

idaniv committed:
Was added tracking of selected elements.

idaniv committed:
Make enabled buttons if selection is correct.

idaniv committed:
Was added icons

idaniv committed:
Remove edit triggers

Guitarheroua commented on src/main.cpp:
```diff
@@ -1,8 +1,19 @@
 {
     QApplication a(argc, argv);
 
+    QStandardItemModel *model = new QStandardItemModel;
```
The model creates a memory leak here because nobody removes the object of it. There is 2 possible solution here: 
1. Give ownership of the model to project viewer window
2. Create a model on the stack, and it removes automatically. 

In future pay attention to this detail, as unnecessary objects will be stored in memory, when you do not need them.

DANIV18 replied:
Ok.So I need to fix it and again send a request?

Guitarheroua replied:
Just fix and push it automatically should update this pull request

Guitarheroua commented on src/main.cpp:
```diff
@@ -1,8 +1,19 @@
+    model->appendRow(new QStandardItem(QIcon("/home/igor/icon"), "path2"));
+    model->appendRow(new QStandardItem(QIcon("/home/igor/icon"), "path3"));
+
+    ProjectViewerWindow *pvw = new ProjectViewerWindow;
```
Same here with ProjectViewerWindow. Create an object on a stack, and it will be removed automatically.

Guitarheroua reviewed (commented):
I run the application, and there are no icons for projects in view. As mentioned in a ticket you should use a white rectangle with text as an icon from the screenshot. This one: 

<img width="131" alt="Знімок екрана  о 12 05 53" src="https://user-images.githubusercontent.com/1516329/60382063-7c0bca80-9a66-11e9-9843-ba8a4eefdd20.png">

idaniv committed:
Was added qrc file with icon

idaniv committed:
delete icon(my little bug) and add correct logo

idaniv committed:
Added correct logo

DANIV18 commented:
@Guitarheroua  try now)

idaniv committed:
Remove one unnecessary declaration

alexchmykhalo commented on src/projectviewerwindow.cpp:
```diff
@@ -0,0 +1,80 @@
+void ProjectViewerWindow::onNewProjectButtonClicked()
+{
+    emit newProjectRequested();
+    qDebug()<<"new project button clicked";
```
Please, remove unnecessary debug output.

DANIV18 replied:
Done.

alexchmykhalo commented on src/projectviewerwindow.cpp:
```diff
@@ -0,0 +1,80 @@
+
+void ProjectViewerWindow::makeButtonsEnabled()
+{
+    qDebug()<<mSelectionModel->selectedIndexes().size();
```
Please, remove unnecessary debug output.

DANIV18 replied:
Done

alexchmykhalo commented on src/projectviewerwindow.cpp:
```diff
@@ -0,0 +1,80 @@
+{
+    qDebug()<<mSelectionModel->selectedIndexes().size();
+    if(mSelectionModel->selectedIndexes().size()== 1)
+    {
```
You do not need so many code to do the same. Open project is enabled if selection count == 1, delete button enabled if selection count > 0. That is 2 bool values that could be set right away.

DANIV18 replied:
However, I still get 2 conditions and one else block.
Like this. Or I'm a fool)
![Screenshot from 2019-06-30 20-46-08](https://user-images.githubusercontent.com/32616149/60400110-24f61a80-9b78-11e9-9575-e52bb303ab48.png)

alexchmykhalo replied:
Could you please assign each condition to separate bool variable. And everything will become clear.

DANIV18 replied:
Done)

alexchmykhalo commented on src/projectviewerwindow.h:
```diff
@@ -0,0 +1,43 @@
+private:
+    Ui::ProjectViewerWindow *ui;
+    QAbstractItemModel *mProjectsViewModel;
+    QItemSelectionModel *mSelectionModel;
```
You do not need to have mSelectionModel as a separate class member. You always have access to view.

DANIV18 replied:
Ok. I've deleted an unnecessary class member.

alexchmykhalo commented on src/projectviewerwindow.h:
```diff
@@ -0,0 +1,43 @@
+class ProjectViewerWindow;
+}
+
+class QString;
```
Please, remove unnecessary forward declarations.

DANIV18 replied:
Done.

alexchmykhalo reviewed (changes requested):
Please, fix review comments.

idaniv committed:
Remove unnecessary debug output.
Remove mSelectionModel as a class member.
Remove unnecessary forward declarations.

idaniv committed:
Complete makeButtonsEnabled method

alexchmykhalo commented on src/projectviewerwindow.h:
```diff
@@ -0,0 +1,38 @@
+
+private:
+    Ui::ProjectViewerWindow *ui;
+    QAbstractItemModel *mProjectsViewModel;
```
Are you sure that you need separate method? Anyway you set this model to the view and always have access to it

alexchmykhalo commented on src/projectviewerwindow.cpp:
```diff
@@ -0,0 +1,67 @@
+    ui->projectsListView->setViewMode(QListView::ViewMode::IconMode);
+    ui->projectsListView->setIconSize(QSize(100,100));
+    ui->projectsListView->setUniformItemSizes(true);
+    ui->projectsListView->setSelectionMode(QAbstractItemView::MultiSelection);
```
Do you really think this mode will be convenient for users? Why users cannot use Ctrl key for multiselection?

DANIV18 replied:
Now users can do this.I used a different selection mode(Extended selection)

alexchmykhalo commented on src/projectviewerwindow.cpp:
```diff
@@ -0,0 +1,67 @@
+#include "projectviewerwindow.h"
+#include "ui_projectviewerwindow.h"
+
+#include <QDebug>
```
Please, remove unused includes

alexchmykhalo commented on src/projectviewerwindow.cpp:
```diff
@@ -0,0 +1,67 @@
+#include "projectviewerwindow.h"
```
Please, move projectviewerwindow.* files to separate folder, add a .pri file that include all that files to project.

DANIV18 replied:
Done. Please, check the project how(maybe I made mistakes again)

idaniv committed:
Set ExtendedSelection as a selection mode.

idaniv committed:
Removed unused includes and class member.

idaniv committed:
Moved projectviewerwindow.* files to separate folder add added .pri file
that include all that files to project.

idaniv committed:
Was added tests for Project Viewer Window.

idaniv committed:
I want to run tests on server.

idaniv committed:
Get appveyor from development branch

idaniv committed:
first try to add correct building

idaniv committed:
Second try

idaniv committed:
Third try

DANIV18 committed:
4

DANIV18 committed:
5

DANIV18 committed:
6

DANIV18 committed:
7

DANIV18 committed:
Finally, it works

alexchmykhalo commented on src/projectviewer/projectviewerwindow.cpp:
```diff
@@ -0,0 +1,57 @@
+    ui->projectsListView->setIconSize(QSize(100,100));
+    ui->projectsListView->setUniformItemSizes(true);
+
+    connect(ui->newProjectButton,&QPushButton::clicked,this,&ProjectViewerWindow::onNewProjectButtonClicked);
```
Please, use spaces after comma for better readability.

alexchmykhalo commented on src/projectviewer/projectviewerwindow.cpp:
```diff
@@ -0,0 +1,57 @@
+
+void ProjectViewerWindow::makeButtonsEnabled()
+{
+    bool isEqualToOne = ui->projectsListView->selectionModel()->selectedIndexes().size() == 1;
```
Please, choose more descriptive names (e g  isOnlyOneProjectSelected, isAnyProjectsSelected)

idaniv committed:
Fixed small issues.
